src/hotspot/share/opto/inlinetypenode.cpp

```diff
@@ -631,6 +631,157 @@ void InlineTypeNode::store(GraphKit* kit, Node* base, Node* ptr, bool immutable_
+    cmp = igvn->register_new_node_with_optimizer(new CmpLNode(val1, val2));
+    break;
+  default:
+    assert(is_reference_type(bt), "must be");
+    cmp = igvn->register_new_node_with_optimizer(new CmpPNode(val1, val2));
+  }
+  Node* bol = igvn->register_new_node_with_optimizer(new BoolNode(cmp, test));
+  IfNode* iff = igvn->register_new_node_with_optimizer(new IfNode(*ctrl, bol, PROB_MAX, COUNT_UNKNOWN))->as_If();
+  Node* if_f = igvn->register_new_node_with_optimizer(new IfFalseNode(iff));
+  Node* if_t = igvn->register_new_node_with_optimizer(new IfTrueNode(iff));
+
+  region->add_req(if_t);
+  if (phi != nullptr) {
+    phi->add_req(igvn->intcon(0));
+  }
+  *ctrl = if_f;
+}
+
+// Check if a substitutability check between 'this' and 'other' can be implemented in IR
+bool InlineTypeNode::can_emit_substitutability_check(Node* other) {
+  if (other != nullptr && other->is_InlineType() && bottom_type() != other->bottom_type()) {
+    // Different types, this is dead code because there's a check above that guarantees this.
+    return false;
+  }
+  for (uint i = 0; i < field_count(); i++) {
+    ciType* ft = field_type(i);
+    if (ft->is_inlinetype()) {
+      // Check recursively
+      if (!field_value(i)->as_InlineType()->can_emit_substitutability_check(nullptr)){
+        return false;
+      }
+    } else if (!ft->is_primitive_type() && ft->as_klass()->can_be_inline_klass()) {
+      // Comparing this field might require (another) substitutability check, bail out
+      return false;
+    }
+  }
+  return true;
+}
+
+// Emit IR to check substitutability between 'this' and the value object referred to by 'ptr' (might be a buffer address or an InlineTypeNode)
+void InlineTypeNode::check_substitutability(PhaseIterGVN* igvn, RegionNode* region, Node* phi, Node** ctrl, Node* mem, Node* base, Node* ptr, bool flat) {
+  BarrierSetC2* bs = BarrierSet::barrier_set()->barrier_set_c2();
+  // TODO 8228361 double-check
+  // DecoratorSet decorators = IN_HEAP | MO_UNORDERED;
+  DecoratorSet decorators = MO_UNORDERED | C2_READ_ACCESS | C2_CONTROL_DEPENDENT_LOAD | IN_HEAP | C2_UNKNOWN_CONTROL_LOAD;
+  MergeMemNode* local_mem = igvn->register_new_node_with_optimizer(MergeMemNode::make(mem))->as_MergeMem();
+
+  ciInlineKlass* vk = inline_klass();
+  for (uint i = 0; i < field_count(); i++) {
+    int field_off = field_offset(i);
+    if (flat) {
+      // Flat access, no header
+      field_off -= vk->payload_offset();
+    }
+    Node* val = field_value(i);
+    ciType* ft = field_type(i);
+    BasicType bt = ft->basic_type();
+
+    Node* field_base = base;
+    Node* field_ptr = ptr;
+
+    // Get field value of other operand
+    if (ptr->is_InlineType()) {
+      field_ptr = ptr->as_InlineType()->field_value(i);
+      field_base = nullptr;
+    } else {
+      assert(base != nullptr, "lost base");
+      field_ptr = igvn->register_new_node_with_optimizer(new AddPNode(base, ptr, igvn->MakeConX(field_off)));
+      if (!field_is_flat(i)) {
```

Comment:
How about adding a `InlineTypeNode::field()` method and then directly calling `is_flat()`, `is_null_free()` on the field instead of reloading it each time over the index?

Reply:
Sounds good, thanks!